packages/contracts-bedrock/src/L2/SoulGasToken.sol

```diff
@@ -151,6 +161,48 @@ contract SoulGasToken is ERC20Upgradeable, OwnableUpgradeable {
         }
     }
 
+    /// @notice allowSgtValue is called by the owner to enable whitelist contracts to consume sgt as msg.value
+    function allowSgtValue(address[] calldata contracts) external onlyOwner {
+        require(IS_BACKED_BY_NATIVE, "allowSgtValue should only be called when IS_BACKED_BY_NATIVE");
+        SoulGasTokenStorage storage $ = _getSoulGasTokenStorage();
+        uint256 i;
+        for (i = 0; i < contracts.length; i++) {
+            $._allowSgtValue[contracts[i]] = true;
+            emit AllowSgtValue(contracts[i]);
+        }
+    }
+
+    /// @notice allowSgtValue is called by the owner to disable whitelist contracts to consume sgt as msg.value
+    function disallowSgtValue(address[] calldata contracts) external onlyOwner {
+        require(IS_BACKED_BY_NATIVE, "disallowSgtValue should only be called when IS_BACKED_BY_NATIVE");
+        SoulGasTokenStorage storage $ = _getSoulGasTokenStorage();
+        uint256 i;
+        for (i = 0; i < contracts.length; i++) {
+            $._allowSgtValue[contracts[i]] = false;
+            emit DisallowSgtValue(contracts[i]);
+        }
+    }
+
+    /// @notice chargeFromOrigin is called when IS_BACKED_BY_NATIVE to charge for native balance
+    ///         from tx.origin if caller is whitelisted.
+    function chargeFromOrigin(uint256 amount) returns (uint256 amountCharged) {
+        require(IS_BACKED_BY_NATIVE, "chargeFromOrigin should only be called when IS_BACKED_BY_NATIVE");
+        SoulGasTokenStorage storage $ = _getSoulGasTokenStorage();
+        require($._allowSgtValue[_msgSender()], "caller is not whitelisted");
+        uint256 balance = balanceOf(tx.origin);
+        if (balance == 0) {
+            amountCharged = 0;
+            return;
+        }
+        if (balance >= amount) {
+            amountCharged = amount;
+        } else {
+            amountCharged = balance;
```

Comment:
It seems a bit odd that if the amount to be charged exceeds the balance, we still attempt to charge it instead of returning 0.

How will the EthStorage contract call this method? Based on my understanding, the EthStorage contract will first check if `msg.value` is sufficient. If not, it will call `chargeFromOrigin(amountNeeded - msg.value)`. If the balance is still insufficient, it will revert.

Reply:
`chargeFromOrigin` may charge less than requested, and the remaining amount is expected to be charged from native balance by the caller.

Reply:
Yes, I prefer charging SGT first in EthStorage and then charging native balance so that it may provider better UX.

Reply:
A new issue has been opened to track the corresponding changes: https://github.com/ethstorage/storage-contracts-v1/issues/122